### PR TITLE
Seed generation and validity

### PR DESCRIPTION
**Pull request by lsolodkova**
Implementation of __generate_seed and __is_valid_seed

### CODE REVIEW

**Review by sobuch — DISMISSED** [review]

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (mvondracek) [inline PA193_mnemonic_Slytherin/mnemonic.py]: What about just
~~~python
return isinstance(seed, bytes) and len(seed) == SEED_LEN
~~~

Can length of seed (`SEED_LEN`) be reconfigured without breaking the BIP39 algorithm?

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (mvondracek) [inline PA193_mnemonic_Slytherin/mnemonic.py]: > The wordlist can contain native characters, but they must be encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD). 
> https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#wordlist

I'm not sure if I understand now, when do you expect UnicodeError to happen?

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (mvondracek) [inline PA193_mnemonic_Slytherin/mnemonic.py]: >we use the PBKDF2 function with a mnemonic sentence (in UTF-8 NFKD) used as the password and the string "mnemonic" + passphrase (again in UTF-8 NFKD) used as the salt
> https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#from-mnemonic-to-seed

I think this is fine, PBKDF2 usually works with *password* and *salt*, therefore we should use mnemonic as the *password* for PBKDF2 and  `'mnemonic' + password_for_ protection_of_the_seed` as *salt*. It's quite confusing. :)

**Review by mvondracek — CHANGES_REQUESTED** [review]

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (lsolodkova) [reply PA193_mnemonic_Slytherin/mnemonic.py]: What do you mean? The length of the seed was taken just 'cause of SHA512. I'm not sure how soon it'll be cracked.

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (lsolodkova) [reply PA193_mnemonic_Slytherin/mnemonic.py]: If the correctness of the input (especially, the password) is guaranteed, then no problem. Not sure about normalizing though, need to think about that if we gonna support non-Latin dictionaries.

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (lsolodkova) [reply PA193_mnemonic_Slytherin/mnemonic.py]: > What about just

Done, thanks.

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (mvondracek) [reply PA193_mnemonic_Slytherin/mnemonic.py]: When we are encoding UTF-8 as bytes, are there posibilities of UnicodeError? We are not trying to get ASCII characters, just values of those bytes.

Comment (mvondracek) [reply PA193_mnemonic_Slytherin/mnemonic.py]: Do we need to be careful about UTF-8 NFKD? Could we have mnemonic in text file which would crash it?

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (mvondracek) [reply PA193_mnemonic_Slytherin/mnemonic.py]: > What do you mean? The length of the seed was taken just 'cause of SHA512. I'm not sure how soon it'll be cracked.

My comment was more about whther SEED_LEN should be global variable or constant in this function. :)

File: PA193_mnemonic_Slytherin/mnemonic.py
Comment (lsolodkova) [reply PA193_mnemonic_Slytherin/mnemonic.py]: Actually, I wasn't able to crash the script, maybe with str to bytes it's nearly impossible to do. But we still have problems if our wordlist file isn't UTF-8 encoded (the same passwords/mnemonics with different encoding produce different results, displaying problem, etc.). That's relevant for non-Latin dictionaries, though.

Comment (mvondracek) [reply PA193_mnemonic_Slytherin/mnemonic.py]: Both mnemonic and dictionary must be UTF-8 according to BIP39. If a dicionary isn't in valid UTF-8, then it's invalid dictionary. Same for mnemonic, if it isn't UTF-8, then we should reject it.

**Review by mvondracek — DISMISSED** [review]
https://github.com/mvondracek/PA193_mnemonic_Slytherin/issues/12#issuecomment-543391802

**Comment by mvondracek** [comment]
I've updated this branch from *dev* so it could be merged without conflicts. 
It can be merged as soon as we have approvals, @lsolodkova, @sobuch.

**Review by sobuch — APPROVED** [review]

**Review by mvondracek — APPROVED** [review]

**Pull request merged**
